File: api/epg.py
import requests
import gzip
import io
from fastapi import FastAPI, HTTPException
from datetime import datetime
import pytz
import xml.etree.ElementTree as ET

app = FastAPI()

def download_and_extract_epg(url):
    response = requests.get(url)
    response.raise_for_status()  # Raise an error for bad responses
    with gzip.GzipFile(fileobj=io.BytesIO(response.content)) as gz:
        xml_contents = gz.read()
    return xml_contents

def convert_to_ist(time_str):
    dt = datetime.strptime(time_str, "%Y%m%d%H%M%S %z")
    ist = dt.astimezone(pytz.timezone('Asia/Kolkata'))
    return ist.strftime('%Y-%m-%d %H:%M:%S')

def get_current_and_upcoming_epg(xml_data, channel_id):
    xml_root = ET.fromstring(xml_data)
    current_time = datetime.now(pytz.timezone('Asia/Kolkata'))

    current_program = None
    next_program = None
    channel_info = None

    # Get channel information
    for channel in xml_root.findall("channel"):  # Make sure to use `findall`
        try:
            if channel.attrib['id'] == channel_id:
                channel_info = {
                    'id': channel.attrib['id'],
                    'name': channel.find("display-name").text,
                    'icon': channel.find("icon").attrib['src']
                }
                break
        except KeyError as e:
            print(f"KeyError: {e} for channel: {ET.tostring(channel)}")
            continue

    if channel_info is None:
        raise HTTPException(status_code=404, detail=f"Channel ID '{channel_id}' not found in EPG.")

    programs = []
    # Get program information
    for programme in xml_root.findall("programme"):
        if programme.attrib['channel'] == channel_id:
            start = convert_to_ist(programme.attrib['start'])
            stop = convert_to_ist(programme.attrib['stop'])
            start_time = datetime.strptime(start, '%Y-%m-%d %H:%M:%S')
            stop_time = datetime.strptime(stop, '%Y-%m-%d %H:%M:%S')

            programs.append({
                'title': programme.find("title").text,
                'desc': programme.find("desc").text,
                'start': start,
                'stop': stop,
                'startTime': start_time,
                'stopTime': stop_time,
                'icon': programme.find("icon").attrib['src']
            })

    # Sort programs by start time
    programs.sort(key=lambda x: x['startTime'])

    # Find current and next program
    for index, program in enumerate(programs):
        if program['startTime'] <= current_time < program['stopTime']:
            current_program = program
            next_program = programs[index + 1] if index + 1 < len(programs) else None
            break

    if not current_program:  # If no current program found, find the next one
        for program in programs:
            if program['startTime'] > current_time:
                next_program = program
                break

    return {
        'Channel': channel_info,
        'Current': current_program,
        'Upcoming': next_program
    }

@app.get("/api/epg")
async def get_epg(id: int):
    channel_id = f'ts{id}'
    url = 'https://avkb.short.gy/tsepg.xml.gz'

    # Download and extract the XML data
    xml_data = download_and_extract_epg(url)

    # Get EPG data for the requested channel
    epg_data = get_current_and_upcoming_epg(xml_data, channel_id)

    return epg_data

File: api/epg.js
// api/epg.js

const axios = require('axios');
const fs = require('fs');
const zlib = require('zlib');
const { parseStringPromise } = require('xml2js');

// Function to download and extract the XML file
async function downloadAndExtractEPG(url) {
    const gzFile = 'epg.xml.gz';
    const xmlFile = 'epg.xml';

    try {
        // Download the gzipped XML file
        const response = await axios.get(url, { responseType: 'arraybuffer' });
        fs.writeFileSync(gzFile, response.data);
        console.log('EPG file downloaded successfully.');
    } catch (error) {
        console.error('Error downloading the EPG:', error);
        throw new Error('Failed to download EPG file');
    }

    try {
        // Extract the gzipped file
        const xmlContents = zlib.gunzipSync(fs.readFileSync(gzFile));
        fs.writeFileSync(xmlFile, xmlContents);
        console.log('EPG file extracted successfully.');
    } catch (error) {
        console.error('Error extracting the EPG:', error);
        throw new Error('Failed to extract EPG file');
    }

    return xmlFile;
}

// Function to convert time to IST
function convertToIST(time) {
    const dateTime = new Date(`${time.slice(0, 8)}T${time.slice(8, 14)}+00:00`);
    return dateTime.toLocaleString('en-IN', { timeZone: 'Asia/Kolkata' });
}

// Function to get the current and upcoming program for a specific channel
async function getCurrentAndUpcomingEPG(xmlFile, channelId) {
    const xmlData = fs.readFileSync(xmlFile);
    const result = await parseStringPromise(xmlData);

    const currentTime = new Date().toLocaleString('en-IN', { timeZone: 'Asia/Kolkata' });
    const programs = [];

    // Get channel information
    const channelInfo = result.tv.channel.find(channel => channel.$.id === channelId);
    if (!channelInfo) {
        return null; // Channel not found
    }

    // Get program information
    for (const programme of result.tv.programme) {
        if (programme.$.channel === channelId) {
            const start = convertToIST(programme.$.start);
            const stop = convertToIST(programme.$.stop);
            programs.push({
                title: programme.title[0],
                desc: programme.desc[0],
                start: start,
                stop: stop,
                icon: programme.icon ? programme.icon[0].$.src : null,
                startTime: new Date(start),
                stopTime: new Date(stop),
            });
        }
    }

    // Sort programs by start time
    programs.sort((a, b) => a.startTime - b.startTime);

    const currentProgram = programs.find(program => program.startTime <= new Date(currentTime) && program.stopTime > new Date(currentTime));
    const nextProgram = currentProgram ? programs[programs.indexOf(currentProgram) + 1] : programs.find(program => program.startTime > new Date(currentTime));

    return {
        Channel: {
            id: channelInfo.$.id,
            name: channelInfo['display-name'][0],
            icon: channelInfo.icon ? channelInfo.icon[0].$.src : null,
        },
        Current: currentProgram,
        Upcoming: nextProgram
    };
}

// Main function to handle requests
module.exports = async (req, res) => {
    const channelId = req.query.id ? 'ts' + req.query.id : null;

    if (!channelId) {
        return res.status(400).json({ error: 'Channel ID is missing.' });
    }

    const url = 'https://avkb.short.gy/tsepg.xml.gz';

    try {
        console.log('Downloading and extracting EPG from URL:', url);
        const xmlFile = await downloadAndExtractEPG(url);
        console.log('EPG file downloaded and extracted:', xmlFile);
        
        const epgData = await getCurrentAndUpcomingEPG(xmlFile, channelId);
        console.log('EPG data retrieved:', epgData);

        // Clean up XML files after processing
        fs.unlinkSync(xmlFile);
        fs.unlinkSync('epg.xml.gz');

        res.setHeader('Content-Type', 'application/json');
        res.status(200).json(epgData);
    } catch (error) {
        console.error('Error occurred:', error);
        res.status(500).json({ error: 'An error occurred while processing the request.' });
    }
};
